File: flutter/flutter-sdk-bridge/ping_journey/android/src/main/kotlin/com/pingidentity/flutter/journey/JourneyConfigParser.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.flutter.journey

import com.pingidentity.flutter.core.CoreRuntime
import com.pingidentity.flutter.core.oidc.OidcConfigHandle
import com.pingidentity.journey.Journey
import com.pingidentity.journey.module.Oidc
import com.pingidentity.logger.Logger
import com.pingidentity.logger.STANDARD
import com.pingidentity.oidc.OidcClientConfig
import com.pingidentity.oidc.OpenIdConfiguration

/**
 * Builds a native [Journey] from the flat, wire-serializable [JourneyConfigMessage].
 *
 * Trimmed to the fields the Pigeon schema carries (no storage/logger handle registries yet —
 * those arrive with a future module). `realm`/`cookie` are left unset when absent so the native
 * SDK's own defaults (`realm = "root"`, `cookie = "iPlanetDirectoryPro"`) apply.
 */
internal object JourneyConfigParser {
    /**
     * Explicit cross-platform default when [JourneyConfigMessage.timeoutMillis] is unset — the
     * native SDKs' own defaults diverge (Android 15s, iOS 30s), which otherwise makes the same
     * unset config wait a different duration per platform with no indication in code or docs.
     */
    const val DEFAULT_TIMEOUT_MILLIS = 30_000L

    fun parse(config: JourneyConfigMessage): Journey =
        Journey {
            serverUrl = config.serverUrl
            config.realm?.let { realm = it }
            config.cookie?.let { cookie = it }
            timeout = config.timeoutMillis ?: DEFAULT_TIMEOUT_MILLIS
            // The SDK's global default is Logger.NONE (a no-op), which also suppresses Ktor's
            // HTTP Logging plugin — every SDK request/response, including the headless OIDC
            // authorize exchange, then produces no logcat output at all. STANDARD routes through
            // android.util.Log under the "Ping SDK <version>" tag. Diagnostic aid, not a
            // production setting: flip to Logger.NONE once the flow is verified.
            logger = Logger.STANDARD

            val oidcClientId = config.oidcClientId
            if (oidcClientId != null) {
                module(Oidc) { applyHandle(resolveOidcHandle(oidcClientId)) }
            } else if (hasOidcConfig(config)) {
                module(Oidc) {
                    config.clientId?.let { clientId = it }
                    config.discoveryEndpoint?.let { discoveryEndpoint = it }
                    config.redirectUri?.let { redirectUri = it }
                    config.scopes?.let { scopes = it.filterNotNull().toMutableSet() }
                    config.acrValues?.let { acrValues = it }
                    config.signOutRedirectUri?.let { signOutRedirectUri = it }
                    config.state?.let { state = it }
                    config.nonce?.let { nonce = it }
                    config.uiLocales?.let { uiLocales = it }
                    config.refreshThreshold?.let { refreshThreshold = it }
                    config.loginHint?.let { loginHint = it }
                    config.display?.let { display = it }
                    config.prompt?.let { prompt = it }
                    config.additionalParameters?.let { params ->
                        additionalParameters =
                            params.entries.filter { it.key != null && it.value != null }
                                .associate { it.key!! to it.value!! }
                    }
                }
            }
        }

    fun hasOidcConfig(config: JourneyConfigMessage): Boolean =
        config.oidcClientId != null ||
            config.clientId != null ||
            config.discoveryEndpoint != null ||
            config.redirectUri != null ||
            config.scopes != null
}

/**
 * Resolves [oidcClientId] from `ping_core`'s shared `CoreRuntime.oidcClientRegistry`, and
 * validates it exposes a usable configuration. Top-level (rather than a member of
 * [JourneyConfigParser]) so unit tests can call it directly — calling it through
 * [JourneyConfigParser.parse] would otherwise require constructing a full native `Journey`, which
 * is not viable in a plain JVM unit test (`Journey.kt`'s own builder reads
 * `android.os.LocaleList.getDefault()` internally, unrelated to this phase, which the stub
 * `android.jar` can't satisfy).
 */
internal fun resolveOidcHandle(oidcClientId: String): OidcConfigHandle {
    val handle =
        CoreRuntime.oidcClientRegistry.resolve(oidcClientId) as? OidcConfigHandle
            ?: throw IllegalArgumentException("OIDC client instance not found for id=$oidcClientId")
    val discoveryEndpoint = handle.discoveryEndpoint?.trim()
    if (discoveryEndpoint.isNullOrEmpty() && handle.openId == null) {
        throw IllegalArgumentException(
            "OIDC client id=$oidcClientId does not expose discoveryEndpoint or openId. " +
                "Configure OIDC with discoveryEndpoint or openId before composing Journey."
        )
    }
    return handle
}

/**
 * Applies every flat field of [handle] — resolved from `ping_core`'s shared
 * `CoreRuntime.oidcClientRegistry` — onto this [OidcClientConfig] receiver, in place of the
 * inline [JourneyConfigMessage] fields. Mirrors `ping_oidc`'s own `OidcConfigParser.applyMessage`
 * (also a top-level extension, for the same reason), so the field list can't drift between the
 * two paths.
 */
internal fun OidcClientConfig.applyHandle(handle: OidcConfigHandle) {
    clientId = handle.clientId
    handle.discoveryEndpoint?.let { discoveryEndpoint = it }
    handle.openId?.let {
        openId =
            OpenIdConfiguration(
                authorizationEndpoint = it.authorizationEndpoint,
                tokenEndpoint = it.tokenEndpoint,
                userinfoEndpoint = it.userinfoEndpoint,
                // Coercions: native SDK's OpenIdConfiguration requires non-null endpoints, but
                // ping_core's OidcOpenIdConfig leaves them nullable (mirrors ping_oidc's own
                // OidcConfigParser.applyMessage).
                endSessionEndpoint = it.endSessionEndpoint ?: "",
                pingEndIdpSessionEndpoint = it.pingEndIdpSessionEndpoint ?: "",
                revocationEndpoint = it.revocationEndpoint ?: "",
            )
    }
    handle.redirectUri?.let { redirectUri = it }
    if (handle.scopes.isNotEmpty()) scopes = handle.scopes.toMutableSet()
    handle.acrValues?.let { acrValues = it }
    handle.signOutRedirectUri?.let { signOutRedirectUri = it }
    handle.state?.let { state = it }
    handle.nonce?.let { nonce = it }
    handle.uiLocales?.let { uiLocales = it }
    handle.refreshThreshold?.let { refreshThreshold = it }
    handle.loginHint?.let { loginHint = it }
    handle.display?.let { display = it }
    handle.prompt?.let { prompt = it }
    if (handle.additionalParameters.isNotEmpty()) additionalParameters = handle.additionalParameters
    par = handle.par
}

File: flutter/flutter-sdk-bridge/ping_journey/ios/ping_journey/Sources/ping_journey/JourneyConfigParser.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingJourney
import PingOidc
import ping_core

/// Builds a native `Journey` from the flat, wire-serializable `JourneyConfigMessage`.
///
/// Trimmed to the fields the Pigeon schema carries (no storage/logger handle registries yet —
/// those arrive with a future module). `realm`/`cookie` are left unset when absent so the native
/// SDK's own defaults (`realm = "root"`, `cookie = "iPlanetDirectoryPro"`) apply.
enum JourneyConfigParser {
    /// Explicit cross-platform default when `JourneyConfigMessage.timeoutMillis` is unset — the
    /// native SDKs' own defaults diverge (Android 15s, iOS 30s), which otherwise makes the same
    /// unset config wait a different duration per platform with no indication in code or docs.
    static let defaultTimeoutMillis: Int64 = 30_000

    static func parse(_ message: JourneyConfigMessage) async throws -> Journey {
        let resolvedHandle: OidcConfigHandle?
        if let oidcClientId = message.oidcClientId {
            resolvedHandle = try await resolveOidcHandle(oidcClientId)
        } else {
            resolvedHandle = nil
        }

        return Journey.createJourney { journeyConfig in
            applyJourneyFields(message, to: journeyConfig)
            if let handle = resolvedHandle {
                journeyConfig.module(PingJourney.OidcModule.config) { oidcConfig in
                    applyHandleFields(handle, to: oidcConfig)
                }
            } else if hasOidcFields(message) {
                journeyConfig.module(PingJourney.OidcModule.config) { oidcConfig in
                    applyOidcFields(message, to: oidcConfig)
                }
            }
        }
    }

    /// Resolves [oidcClientId] from `ping_core`'s shared `CoreRuntime.oidcClientRegistry`. Not
    /// `private` so `@testable import` can exercise resolution/validation directly, without
    /// going through `parse`'s full `Journey.createJourney`.
    static func resolveOidcHandle(_ oidcClientId: String) async throws -> OidcConfigHandle {
        guard let handle = await CoreRuntime.oidcClientRegistry.resolve(oidcClientId) as? OidcConfigHandle else {
            throw JourneyHostApiError.argument("OIDC client instance not found for id=\(oidcClientId)")
        }
        let discoveryEndpoint = handle.discoveryEndpoint?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !(discoveryEndpoint == nil || discoveryEndpoint!.isEmpty) || handle.openId != nil else {
            throw JourneyHostApiError.argument(
                "OIDC client id=\(oidcClientId) does not expose discoveryEndpoint or openId. " +
                "Configure OIDC with discoveryEndpoint or openId before composing Journey."
            )
        }
        return handle
    }

    /// Applies every flat field of [handle] — resolved from `ping_core`'s shared
    /// `CoreRuntime.oidcClientRegistry` — onto [oidcConfig], in place of the inline
    /// `JourneyConfigMessage` fields. Mirrors `ping_oidc`'s own `OidcConfigParser.apply`, so the
    /// field list can't drift between the two paths. Not `private`, for the same
    /// direct-testability reason as `resolveOidcHandle`.
    static func applyHandleFields(_ handle: OidcConfigHandle, to oidcConfig: OidcClientConfig) {
        oidcConfig.clientId = handle.clientId
        if let discoveryEndpoint = handle.discoveryEndpoint { oidcConfig.discoveryEndpoint = discoveryEndpoint }
        if let openId = handle.openId {
            // openId can only be applied via openIdOverride — OidcClientConfig.openId is
            // `public private(set)` at ping-ios-sdk 2.1.0 (SDKS-5301 tracks a cleaner API). This
            // closure only fires after discover() succeeds, which requires discoveryEndpoint —
            // ping_oidc's own OidcConfigParser already guarantees any handle registered via
            // OidcClient.configure() has one, so this is safe here without re-validating.
            oidcConfig.openIdOverride = { discovered in
                discovered.authorizationEndpoint = openId.authorizationEndpoint
                discovered.tokenEndpoint = openId.tokenEndpoint
                discovered.userinfoEndpoint = openId.userinfoEndpoint
                discovered.endSessionEndpoint = openId.endSessionEndpoint ?? ""
                discovered.revocationEndpoint = openId.revocationEndpoint ?? ""
                discovered.pingEndsessionEndpoint = openId.pingEndIdpSessionEndpoint
            }
        }
        if let redirectUri = handle.redirectUri { oidcConfig.redirectUri = redirectUri }
        if !handle.scopes.isEmpty { oidcConfig.scopes = Set(handle.scopes) }
        if let acrValues = handle.acrValues { oidcConfig.acrValues = acrValues }
        // signOutRedirectUri: no equivalent field on iOS's OidcClientConfig at 2.1.0 — same
        // asymmetry already documented for the existing inline path
        // (ping_oidc/ios/.../OidcConfigParser.swift).
        if let state = handle.state { oidcConfig.state = state }
        if let nonce = handle.nonce { oidcConfig.nonce = nonce }
        if let uiLocales = handle.uiLocales { oidcConfig.uiLocales = uiLocales }
        if let refreshThreshold = handle.refreshThreshold { oidcConfig.refreshThreshold = refreshThreshold }
        if let loginHint = handle.loginHint { oidcConfig.loginHint = loginHint }
        if let display = handle.display { oidcConfig.display = display }
        if let prompt = handle.prompt { oidcConfig.prompt = prompt }
        if !handle.additionalParameters.isEmpty { oidcConfig.additionalParameters = handle.additionalParameters }
        oidcConfig.par = handle.par
    }

    private static func applyJourneyFields(_ message: JourneyConfigMessage, to journeyConfig: JourneyConfig) {
        journeyConfig.serverUrl = message.serverUrl
        if let realm = message.realm { journeyConfig.realm = realm }
        if let cookie = message.cookie { journeyConfig.cookie = cookie }
        let timeoutMillis = message.timeoutMillis ?? defaultTimeoutMillis
        journeyConfig.timeout = TimeInterval(timeoutMillis) / 1000
    }

    private static func applyOidcFields(_ message: JourneyConfigMessage, to oidcConfig: OidcClientConfig) {
        if let clientId = message.clientId { oidcConfig.clientId = clientId }
        if let discoveryEndpoint = message.discoveryEndpoint {
            oidcConfig.discoveryEndpoint = discoveryEndpoint
        }
        if let redirectUri = message.redirectUri { oidcConfig.redirectUri = redirectUri }
        if let scopes = message.scopes {
            oidcConfig.scopes = Set(scopes.compactMap { $0 })
        }
        if let acrValues = message.acrValues { oidcConfig.acrValues = acrValues }
        if let state = message.state { oidcConfig.state = state }
        if let nonce = message.nonce { oidcConfig.nonce = nonce }
        if let uiLocales = message.uiLocales { oidcConfig.uiLocales = uiLocales }
        if let refreshThreshold = message.refreshThreshold {
            oidcConfig.refreshThreshold = refreshThreshold
        }
        if let loginHint = message.loginHint { oidcConfig.loginHint = loginHint }
        if let display = message.display { oidcConfig.display = display }
        if let prompt = message.prompt { oidcConfig.prompt = prompt }
        if let additionalParameters = message.additionalParameters {
            oidcConfig.additionalParameters = additionalParameters.reduce(into: [String: String]()) { result, entry in
                if let key = entry.key, let value = entry.value {
                    result[key] = value
                }
            }
        }
    }

    static func hasOidcFields(_ message: JourneyConfigMessage) -> Bool {
        message.oidcClientId != nil
            || message.clientId != nil
            || message.discoveryEndpoint != nil
            || message.redirectUri != nil
            || message.scopes != nil
    }
}
